src/Mod/Sketcher/App/Sketch.cpp

```diff
@@ -1743,6 +1743,28 @@ int Sketch::addConstraint(const Constraint *constraint)
                 }
             }
         }
+        else if (constraint->FirstPos == PointPos::start &&
+                 (constraint->SecondPos == PointPos::start ||
+                  constraint->SecondPos == PointPos::end) &&
+                 constraint->Third == GeoEnum::GeoUndef) {
+            // check for B-Spline Knot to curve tangency
+            auto knotgeoId = checkGeoId(constraint->First);
+            if (Geoms[knotgeoId].type == Point) {
+                auto *point = static_cast<const GeomPoint*>(Geoms[knotgeoId].geo);
+
+                if (GeometryFacade::isInternalType(point,InternalType::BSplineKnotPoint)) {
+                    auto bsplinegeoid = internalAlignmentGeometryMap.at(constraint->First);
+
+                    bsplinegeoid = checkGeoId(bsplinegeoid);
+
+                    auto linegeoid = checkGeoId(constraint->Second);
+
+                    if (Geoms[linegeoid].type == Line)
+                        rtn = addTangentLineEndpointAtBSplineKnotConstraint(
+                            linegeoid, constraint->SecondPos, bsplinegeoid, knotgeoId);
+                }
+            }
+        }

```

Comment:
I am just reading the code, but isn't this block exactly the same as the previous block, save for, in the "if" checking constraint->SecondPos for none, start, end, and  calling a different function?

It seems it could be refactored into:
```
else if (constraint->FirstPos == PointPos::start &&
                 constraint->SecondPos != PointPos::mid &&
                 constraint->Third == GeoEnum::GeoUndef) {
            // check for B-Spline Knot to curve tangency
            auto knotgeoId = checkGeoId(constraint->First);
            if (Geoms[knotgeoId].type == Point) {
                auto *point = static_cast<const GeomPoint*>(Geoms[knotgeoId].geo);

                if (GeometryFacade::isInternalType(point,InternalType::BSplineKnotPoint)) {
                    auto bsplinegeoid = internalAlignmentGeometryMap.at(constraint->First);

                    bsplinegeoid = checkGeoId(bsplinegeoid);

                    auto linegeoid = checkGeoId(constraint->Second);

                    if (Geoms[linegeoid].type == Line) {
                        if(constraint->SecondPos == PointPos::none) {
                            rtn = addTangentLineAtBSplineKnotConstraint(
                                linegeoid, bsplinegeoid, knotgeoId);
                        }
                        else { // constraint->SecondPos == PointPos::start || constraint->SecondPos == PointPos::end
                            rtn = addTangentLineEndpointAtBSplineKnotConstraint(
                                linegeoid, constraint->SecondPos, bsplinegeoid, knotgeoId);
                       }
                   }
                }
            }
        }
```

Does it make sense?

Reply:
This is kind of what I meant by the logic getting convoluted. The more support we add, the more special cases, and the more need for streamlining...